### PR TITLE
kate's tcp chat server, complete sans chai tests

### PR DESCRIPTION
**Pull request by katesowles**

### CODE REVIEW

**Comment by katesowles** [comment]
resubmitted WITH two chai tests.

File: chat.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+const random = require('./random');
+
+chat.portNum = 6500;
+chat.sockets = [];
+
+chat.server = net.createServer(socket => {
+
+  socket.username = random();
+  var username = socket.username;
+
```
Comment (martypdx) [inline chat.js]: fyi, you can write this as:
```js
var username = socket.username = random();
```

File: chat.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    chat.broadcast(username, message);
+  });
+
+  socket.on('error', error => {
+    console.log(`Error: ${error.message}`);
+  });
+});
+
+chat.remove = function(socket) {
+  chat.sockets.splice(chat.sockets.indexOf(socket.username), 1);
```
Comment (martypdx) [inline chat.js]: Be careful with indexOf in splice because if not found it returns -1 which splices from end

File: chat.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  if (chat.sockets.length === 0) {
+    process.stdout.write('everyone left the chat');
+    return;
+  }
+
+  chat.sockets.forEach(function(socket) {
+    if (socket.username === from) return;
+    socket.write(message);
+  });
+};
```
Comment (martypdx) [inline chat.js]: `from` is technically javascript reserved word, meaning it _may_ be added in future js. Generally works but syntax highlight is messed up.

**Pull request merged**